### PR TITLE
Initial database support for persisting results.

### PR DESCRIPTION
**Pull request by cpu**
Three database tables are specified in the provided `db-schema.sql`:
1) Experiments - holding an ID for each experiment, the start time, the end time, and the `dnslol` command line arguments used.
2) Servers - holding an ID for each individual DNS server address, an ID for the corresponding Experiment, and the hostname of each DNS server. This table allows an experiment to have many servers.
3) Results - holding an ID for each individual query result, the name that was queried, the type of query that was performed, the recorded error (if any), the ID of the server that was queried, and the ID of the experiment that the query was associated with.

`dnslol` is updated to accept a MySQL DSN to specify the username/password/host/databasename for the DB connection. When an Experiment is started the corresponding Experiments and Servers rows are
created. Results are written to the Results table as they are received.

A new `-count` argument is added to `dnslol` to allow doing `count` copies of each query per nameserver within one experiment. This will make it easier to find flaky names.

After discovering some unicode IDN domains snuck into the dataset (causing DB persist errors due to MySQL `utf8` vs `utf8mb4`) this branch also updates `dnslol` to reject non-ascii domains. The README is updated to describe this and to provide a small Go program that can convert IDN to IDNA2008 encoded ascii.

### CODE REVIEW

File: dnslol/dnslol.go [hunk 1]
```diff
@@ -134,50 +152,72 @@ func (e Experiment) runQueries(dnsClient *dns.Client, name string) error {
...
 			resultLabels["result"] = err.Error()
 		} else {
 			// If the result was successful, increment the success stat and put
 			// "ok" in the result label
-			stats.successes.With(prom.Labels{"server": q.Server}).Add(1)
+			stats.successes.With(prom.Labels{"server": q.Server.address}).Add(1)
 			resultLabels["result"] = "ok"
-			outcome = "ok"
 		}
 		// TODO(@cpu): This should be a separate function.
```
Comment (jsha) [inline dnslol/dnslol.go]: Remove this TODO? :-D

File: go.mod [hunk 2]
```diff
@@ -12,4 +12,5 @@ require (
 	golang.org/x/crypto v0.0.0-20181015023909-0c41d7ab0a0e // indirect
 	golang.org/x/net v0.0.0-20181023162649-9b4f9f5ad519 // indirect
 	golang.org/x/sys v0.0.0-20181023152157-44b849a8bc13 // indirect
+	google.golang.org/appengine v1.2.0 // indirect
```
Comment (jsha) [inline go.mod:15]: Why did we wind up with an appengine dependency?

Comment (cpu) [reply go.mod:15]: `go mod vendor` is pulling it in as an indirect dependency from `github.com/go-sql-driver/mysql`

Comment (cpu) [reply go.mod:15]: I think `go mod vendor` might not be smart enough to know about this build tag?

https://github.com/go-sql-driver/mysql/blob/749ddf1598b47e3cd909414bda735fe790ef3d30/appengine.go#L9

Comment (cpu) [reply go.mod:15]: Looks like this issue: https://github.com/golang/go/issues/26968 - the exclude tags arg for `go mod tidy` mentioned in the last comment & linked issue aren't implemented/decided on yet.

File: dnslol/dnslol.go (hunk 1 above)
Comment (cpu) [reply dnslol/dnslol.go]: Good catch :laughing: fixed in 9e56c9e

File: dnslol/dnslol.go [hunk 3]
```diff
@@ -205,18 +253,16 @@ func (e Experiment) buildQueries(name string) []query {
 // other than RcodeSuccess, a false bool and a nil error or an empty rcode
 // string, a false bool and a non-nil error. In all cases the queryTimes latency
 // stat is updated for the server and query type performed.
-//
-// TODO(@cpu): Update comment
```
Comment (jsha) [inline dnslol/dnslol.go:209]: I think this comment still needs updating.

File: dnslol/dnslol.go [hunk 4]
```diff
@@ -256,13 +372,30 @@ func Start(e Experiment, names <-chan string, wg *sync.WaitGroup) error {
...
 	}()
 
+	// Connect to the database
+	db, err := sql.Open("mysql", dsn)
+	if err != nil {
+		return err
+	}
+	db.SetMaxOpenConns(maxConns)
+	// Reuse connections forever
+	db.SetConnMaxLifetime(0)
```
Comment (jsha) [inline dnslol/dnslol.go]: I thought you were going to remove this line?

**Review by jsha — DISMISSED** [review]

File: dnslol/dnslol.go (hunk 3 above)
Comment (cpu) [reply dnslol/dnslol.go:209]: Oops, good catch. Thanks!

Comment (cpu) [reply dnslol/dnslol.go:209]: Updated in 526186a

File: dnslol/dnslol.go (hunk 4 above)
Comment (cpu) [reply dnslol/dnslol.go]: Yup, I removed it on the copy of the tool I was using on a Droplet and forgot to do it in my git source tree locally. D'oh.

Removed in 526186a

**Review by jsha — APPROVED** [review]

**Pull request merged**